guides/micronaut-security-x509/micronaut-security-x509.adoc

```diff
@@ -0,0 +1,444 @@
+        -inkey client.key -in client.pem
+----
+
+You will be asked for a new password for the PKCS #12 archive:
+
+image::x509/certs9.png[]
+
+This will create `client.p12`. You can verify and inspect the archive with
+
+[source,bash]
+----
+openssl pkcs12 -in client.p12 -noout -info
+----
+
+== Writing the Application
+
+common:cli-or-launch.adoc[]
+
+[source,bash]
+----
+mn create-app example.micronaut.micronautguide \
+   --features=@features@ \
+   --build=@build@ --lang=@lang@
+----
+
+common:build-lang-arguments.adoc[]
+
+common:default-package.adoc[]
+
+diffLink:[]
+
+common:annotationprocessors.adoc[]
+
+=== Copy Certificates
+
+Create the `src/main/resources/ssl` directory and copy the keystore (`keystore.p12`) and truststore (`truststore.jks`) that you created earlier into it.
+
+=== Configuration
+
+Replace the generated `application.yml` with this:
+
+resource:application.yml[]
+
+<1> Enable X.509 support
+<2> Enable SSL/TLS at the server
+<3> Set `client-authentication` to `want`, i.e. accept a client certificate if available, but don't block access without one. To require a client certificate for all requests, change to `need`
+<4> Specify the location of the keystore
+<5> Set this value with the password you used when creating the keystore
+<6> Specify the keystore type as PKCS #12
+<7> Specify the location of the truststore
+<8> Set this value with the password you used when creating the truststore
+<9> Specify the truststore type as JKS (Java Key Store)
+
+=== HelloController
+
+Create `HelloController` to test using client X.509 certificates:
+
+source:HelloController[]
+
+<1> Allow authenticated and non-authenticated requests
+callout:text-plain[2]
+<3> Bind the `X509Authentication` if present, but allow null when not authenticated
+<4> Redundantly bind the `Authentication` to verify that it's the same instance as the `X509Authentication`
+<5> Default the username for non-authenticated requests
+<6> Sanity check that there's no `Authentication` if there's no `X509Authentication`
+<7> Sanity check that the `Authentication` is the `X509Authentication`
+<8> Build the response as the username plus info about the certificate CA for the authenticated user
+
+=== Writing Tests
+
+Create a test to verify that X.509 authentication works:
+
+test:X509Test[]
+
+callout:micronaut-test[1]
+callout:http-client[2]
+<3> Make a request to the controller
+<4> Verify the response is correct
+
+Create `application-test.yml` in `src/test/resources` with this content:
+
+testResource:application-test.yml[tag=client]
+
+<1> Start the server on a random unused SSL port
+<2> Specify the location of the client keystore
+<3> Change this value to the password you used when creating the client keystore
+<4> Specify the keystore type as PKCS #12
+
+common:testApp.adoc[]
+
+== Running the Application
+
+:exclude-for-build:maven
+
+To run the application, use the `./gradlew run` command which starts the application on port 8443.
```

Comment:
```suggestion
To run the application, use the `./gradlew run` command, which starts the application on port 8443.
```
This sentence could have more than one meaning, and I'm not sure which one is correct.

The edit above says that the ./gradelw run command is the command that one should use, and it will start the application on port 8443.

Alternately, if there are multiple ./gradlew run commands, and you want the reader to choose the one that starts the application on port 8443 (as opposed to the one that starts the app on a different port), it should be written as

```suggestion
To run the application, use the `./gradlew run` command that starts the application on port 8443.
```